subprojects/model-core/src/test/groovy/org/gradle/internal/instantiation/generator/DependencyInjectionUsingLenientConstructorSelectorTest.groovy

```diff
@@ -124,14 +125,31 @@ class DependencyInjectionUsingLenientConstructorSelectorTest extends Specificati
         e.cause.message == "No constructors of type DependencyInjectionUsingLenientConstructorSelectorTest.HasConstructors match parameters: ['a', 'b']"
     }
 
-    def "fails when no constructors are ambiguous"() {
+    def "fails when constructors are ambiguous"() {
         when:
         instantiator.newInstance(HasConstructors, ["a"] as Object[])
 
         then:
         ObjectInstantiationException e = thrown()
         e.cause instanceof IllegalArgumentException
-        e.cause.message == "Multiple constructors of type DependencyInjectionUsingLenientConstructorSelectorTest.HasConstructors match parameters: ['a']"
+        TextUtil.normaliseLineSeparators(e.cause.message) == """Multiple constructors for parameters ['a']:
+  1. candidate: DependencyInjectionUsingLenientConstructorSelectorTest.HasConstructors(String, Number)
+  2. best match: DependencyInjectionUsingLenientConstructorSelectorTest.HasConstructors(String, boolean)"""
+    }
+
+    def "uses exact match constructor when constructors could be ambiguous with multiple parameters"() {
+        when:
+        // HasSeveralParameters has 3 constructors
+        // 1 with 3 parameters and 2 with 4 parameters
+        // the constructor with 3 parameters is a subset of the 4 parameter constructor
+        def inst = instantiator.newInstance(HasSeveralParameters, ["a", 0, false] as Object[])
+
+        then:
+        inst.a == "a"
+        inst.b == 0
+        inst.c == 99
+        !inst.d
+        inst.x == "A B D"
     }
```

Comment:
This is a good test, but took some examination of the `HasSeveralParameters` helper to understand what it's doing.

Should we also unit test the `ConstructorComparator`? For example, make two POJOs each having different constructors, then compare their values by using the comparator directly?

Reply:
Same goes for `ManyConstructors` in `AbstractClassGeneratorSpec`; it tests the end-result that constructors are deterministically sorted by number of arguments - but a simpler test of the comparator might be nice too.

Reply:
PTAL again, I changed the way we sense which constructor was used and added a unit test for the comparator